### ISSUE TITLE
`path_statements` doesn't realise that dropping a place expression can assert validity

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#120857]
I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e352e6d248a7b3691c0163c384bb40b9)):

```rust
#![allow(unused_variables)]

#[repr(C)]
union Bar {
    zero: u8,
    nonzero: std::num::NonZeroU8,
}

fn main() {
    unsafe {
        match (Bar { zero: 0 }) {
            Bar { nonzero } if false => (),
            Bar { nonzero } if { nonzero; false } => (),
            _ => (),
        }
    }
}
```

I expected to see this happen: no warning

Instead, this happened:

```
warning: path statement with no effect
  --> src/main.rs:13:34
   |
13 |             Bar { nonzero } if { nonzero; false } => (),
   |                                  ^^^^^^^^
   |
   = note: `#[warn(path_statements)]` on by default
```

When executed under Miri you can see that this path-statement _has_ an effect, causing UB when the second guard is executed, as it causes a typed-copy out of the place-expression unlike the first guard:

```
error: Undefined Behavior: constructing invalid value: encountered 0, but expected something greater or equal to 1
  --> src/main.rs:13:34
   |
13 |             Bar { nonzero } if { nonzero; false } => (),
   |                                  ^^^^^^^ constructing invalid value: encountered 0, but expected something greater or equal to 1
   |
```

### Meta

`rustc --version --verbose`:
```
2024-02-08 98aa3624be70462d6a25
```

**Comment by RalfJung** [comment]
> When executed under Miri you can see that this path-statement has an effect, causing UB when the second guard is executed, as it causes a typed-copy out of the place-expression unlike the first guard:

However, even if it has UB, then removing it is a semantics-preserving transformation. In that sense I think saying "no effect" is justified: you can never introduce a bug or cause a program to violate its specification by removing this statement.

**Comment by Nilstrieb** [comment]
If you're using a path statement to assert validity, you definitely need to put a comment there explaining wtf you're doing,  and an `allow` helps too.
I see no reason for rustc to not lint based on this.

**Comment by RalfJung** [comment]
@rust-lang/opsem  what do you think? I think we should close this. "no effect" here means "is refined by a NOP", which is not the same as "is equivalent to a NOP". It's probably too confusing to explain that in this lint though.

**Comment by Noratrieb** [comment]
It looks like T-opsem does not think, but I do, and I think it should be closed.
